### ISSUE TITLE
[studio] Publisher breaks when moving folder, updating content inside folder, staging content and then publish everything

### ISSUE THREAD

**Issue by Euquimides** [craftercms/craftercms#6373]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [x] The issue is in the develop branch
- [X] The issue is in the latest released 4.1.x
- [ ] The issue is in the latest released 4.0.x
- [ ] The issue is in the latest released 3.1.x

### Describe the issue

Publisher breaks when moving folder, updating content inside folder, staging content and then publish everything. This sets the publishing repository to a `staging_in_progress` state. 

### Steps to reproduce

Steps:

1. Create a new site.
2. Make a first publish.
3. Modify the internal name of a page, for example Health --> Healthsss.
4. Configure the site to include staging.
5. Create a new folder named `test`.
6. Copy and paste Healthsss into the new folder.
7. Got to `Project Tools` --> `Publishing` --> `Publish All` to live.
8. Cut and paste the folder with its contents inside `Healthsss`
9. Modify the internal name of the page inside the folder.
10. Publish the modified page to `Staging`
11. `Publish All` to live again.
12. See the issue

### Relevant log output

```shell
[INFO] 2023-12-15T12:47:57,299 [http-nio-8080-exec-3] [deployment.DeploymentServiceImpl] | Publishing started for site 'a2' 
[INFO] 2023-12-15T12:47:57,306 [http-nio-8080-exec-3] [deployment.DeploymentServiceImpl] | Publishing started for site 'a2' 
[INFO] 2023-12-15T12:47:57,577 [studio.clockTaskExecutor-29] [job.StudioPublisherTask] | Publish started in site 'a2' for target 'staging' with '2' items ready to be published 
[INFO] 2023-12-15T12:47:57,583 [studio.clockTaskExecutor-29] [job.StudioPublisherTask] | Publish '2' items in site 'a2' to target 'staging' 
[ERROR] 2023-12-15T12:47:57,733 [studio.clockTaskExecutor-29] [repository.GitContentRepository] | Failed to publish site 'a2' to publishing target 'staging' commit ID is '8b0409ba57b3dc23bf5492c77a8b28f83807cad4' 
java.lang.NullPointerException: Cannot invoke "java.nio.file.Path.toString()" because "parentFolder" is null
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.deleteParentFolder(GitContentRepository.java:1106) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1097) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.publish(GitContentRepository.java:942) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.blob.BlobAwareContentRepository.publish(BlobAwareContentRepository.java:583) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.deploy(StudioPublisherTask.java:299) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.doPublishing(StudioPublisherTask.java:223) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.executeInternal(StudioPublisherTask.java:130) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioClockTask.execute(StudioClockTask.java:68) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioClockExecutor.lambda$executeTasks$0(StudioClockExecutor.java:92) ~[classes/:4.1.3-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:840) [?:?]
[ERROR] 2023-12-15T12:47:57,733 [studio.clockTaskExecutor-29] [blob.BlobAwareContentRepository] | Failed to publish items in site 'a2' to target 'staging' 
org.craftercms.studio.api.v1.service.deployment.DeploymentException: Failed to publish site 'a2' to publishing target 'staging' commit ID is '8b0409ba57b3dc23bf5492c77a8b28f83807cad4'
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.publish(GitContentRepository.java:1056) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.blob.BlobAwareContentRepository.publish(BlobAwareContentRepository.java:583) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.deploy(StudioPublisherTask.java:299) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.doPublishing(StudioPublisherTask.java:223) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.executeInternal(StudioPublisherTask.java:130) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioClockTask.execute(StudioClockTask.java:68) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioClockExecutor.lambda$executeTasks$0(StudioClockExecutor.java:92) ~[classes/:4.1.3-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: java.lang.NullPointerException: Cannot invoke "java.nio.file.Path.toString()" because "parentFolder" is null
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.deleteParentFolder(GitContentRepository.java:1106) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1097) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.publish(GitContentRepository.java:942) ~[classes/:4.1.3-SNAPSHOT]
	... 9 more
[ERROR] 2023-12-15T12:47:57,734 [studio.clockTaskExecutor-29] [job.StudioPublisherTask] | Failed to publish '2' items in site 'a2' 
org.craftercms.studio.api.v1.service.deployment.DeploymentException: Failed to publish items in site 'a2' to target 'staging'
	at org.craftercms.studio.impl.v2.repository.blob.BlobAwareContentRepository.publish(BlobAwareContentRepository.java:586) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.deploy(StudioPublisherTask.java:299) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.doPublishing(StudioPublisherTask.java:223) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.executeInternal(StudioPublisherTask.java:130) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioClockTask.execute(StudioClockTask.java:68) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioClockExecutor.lambda$executeTasks$0(StudioClockExecutor.java:92) ~[classes/:4.1.3-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: org.craftercms.studio.api.v1.service.deployment.DeploymentException: Failed to publish site 'a2' to publishing target 'staging' commit ID is '8b0409ba57b3dc23bf5492c77a8b28f83807cad4'
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.publish(GitContentRepository.java:1056) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.blob.BlobAwareContentRepository.publish(BlobAwareContentRepository.java:583) ~[classes/:4.1.3-SNAPSHOT]
	... 8 more
Caused by: java.lang.NullPointerException: Cannot invoke "java.nio.file.Path.toString()" because "parentFolder" is null
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.deleteParentFolder(GitContentRepository.java:1106) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1097) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.publish(GitContentRepository.java:942) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.blob.BlobAwareContentRepository.publish(BlobAwareContentRepository.java:583) ~[classes/:4.1.3-SNAPSHOT]
	... 8 more
[ERROR] 2023-12-15T12:47:57,758 [studio.clockTaskExecutor-29] [job.StudioPublisherTask] | Failed to publish items in site 'a2' 
org.craftercms.studio.api.v1.service.deployment.DeploymentException: Failed to publish items in site 'a2' to target 'staging'
	at org.craftercms.studio.impl.v2.repository.blob.BlobAwareContentRepository.publish(BlobAwareContentRepository.java:586) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.deploy(StudioPublisherTask.java:299) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.doPublishing(StudioPublisherTask.java:223) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioPublisherTask.executeInternal(StudioPublisherTask.java:130) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioClockTask.execute(StudioClockTask.java:68) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.job.StudioClockExecutor.lambda$executeTasks$0(StudioClockExecutor.java:92) ~[classes/:4.1.3-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: org.craftercms.studio.api.v1.service.deployment.DeploymentException: Failed to publish site 'a2' to publishing target 'staging' commit ID is '8b0409ba57b3dc23bf5492c77a8b28f83807cad4'
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.publish(GitContentRepository.java:1056) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.blob.BlobAwareContentRepository.publish(BlobAwareContentRepository.java:583) ~[classes/:4.1.3-SNAPSHOT]
	... 8 more
Caused by: java.lang.NullPointerException: Cannot invoke "java.nio.file.Path.toString()" because "parentFolder" is null
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.deleteParentFolder(GitContentRepository.java:1106) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1097) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.cleanUpMoveFolders(GitContentRepository.java:1101) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.GitContentRepository.publish(GitContentRepository.java:942) ~[classes/:4.1.3-SNAPSHOT]
	at org.craftercms.studio.impl.v2.repository.blob.BlobAwareContentRepository.publish(BlobAwareContentRepository.java:583) ~[classes/:4.1.3-SNAPSHOT]
	... 8 more
```

### Screenshots and/or videos

![Screenshot from 2023-12-15 12-57-57](https://github.com/craftercms/craftercms/assets/90878288/9ccdd556-fb30-4980-ba31-b82cbd64823a)

**Comment by sumerjabri** [comment]
@jmendeza we're not sure if this is in 4.0 or not, but it's not in 4.2. Please fix in 4.1 and in 4.0 if there.

**Comment by sumerjabri** [comment]
The fix approach:
- If `staging` is enabled, and there is no `staging` branch, then
  - if `live` branch is present, clone `staging` from `live`
  - else, create a clean `staging`
- `PublishAll` must honor the `staging` configuration

Longer-term, we will consider enhancements after we refactor the publisher in 4.2.

**Comment by jmendeza** [comment]
> The fix approach:
> 
> * If `staging` is enabled, and there is no `staging` branch, then
>   
>   * if `live` branch is present, clone `staging` from `live`
>   * else, create a clean `staging`
> * `PublishAll` must honor the `staging` configuration
> 
> Longer-term, we will consider enhancements after we refactor the publisher in 4.2.

When publishing to `staging` and neither `staging` nor `live` branches exist: if `live` does not exist, then this is the initial publish and we should publish everything. Then `staging` should be created from `sandbox`.

@sumerjabri

**Comment by jmendeza** [comment]
Even though this exact behavior is not reproducible in 4.2, there is still an issue when enabling staging after initial publish.
In that scenario, when publishing content after enabling staging, staging branch might contain just the most recently published content.

**Comment by Euquimides** [comment]
Hi! Updated and rechecked this issue. Fix verified for Crafter v4.0.8, v4.1.3 and v4.2.0. Thank you! Closing this ticket.
